### PR TITLE
Add CLI commands for nonces

### LINKED ISSUE #7118: CLI - Support creating a new nonce account for offline tx signing
#### Problem
We currently don't have an easy way to create an initialized nonce account.

Signing a transaction offline is slow and may require passing a `nonce` (valid blockhash) which will be verified by the nonce program. In order to consume the nonce, it must already be initialized in a separate transaction. Then, the nonce account private key must be used to sign the nonced transaction. 

#### Proposed Solution
Add a CLI command for creating and initializing a nonce account

`solana create-nonce-account ~/.config/solana/nonce-keypair.json`

Supported options:
- `--sign-only` for offline signing
- `--signatures` to use offline signature
- `ASK` instead of keypair path for secure key input

### PR DESCRIPTION
**Pull request by jackcmay**
#### Problem

No support for nonces in the CLI

#### Summary of Changes

Added:
- `create-nonce-account`
- `get-nonce`
- `new-nonce`
- `show-nonce-account`
- `withdraw-from-nonce-account`

Depends on #6845

Fixes #7118

### CODE REVIEW

**Comment by jackcmay** [comment]
@t-nelson fyi

**Comment by t-nelson** [comment]
Awesome! Thanks!

I'll make sure we get this in between #6845 and the implementation of #7305

**Comment by jackcmay** [comment]
> Awesome! Thanks!
> 
> I'll make sure we get this in between #6845 and the implementation of #7305

Sounds good, I'll follow up with any CLI changes that are necessary

**Comment by t-nelson** [comment]
Tag, you're it!

File: cli/src/cli.rs [hunk 1]
```diff
@@ -206,6 +207,22 @@ pub enum CliCommand {
     },
     TimeElapsed(Pubkey, Pubkey, DateTime<Utc>), // TimeElapsed(to, process_id, timestamp)
     Witness(Pubkey, Pubkey),                    // Witness(to, process_id)
+    // Nonce commands
+    CreateNonceAccount {
+        nonce_account: KeypairEq,
+        lamports: u64,
+    },
+    NewNonce(KeypairEq),
+    GetNonce(Pubkey),
```
Comment (CriesofCarrots) [inline cli/src/cli.rs]: Some A-Z requests... 
(1) Could we alphabetize the Nonce commands? I know it makes sense to put them in usage order, and we used to do a lot of that here, but it made it pretty difficult to find anything in this ginormous file.
(2) Could you move the "Nonce commands" block above "Program Deployment"? (and same request for the `parse_` match block and the `process_` match block)

File: cli/src/nonce.rs [hunk 2]
```diff
@@ -0,0 +1,466 @@
...
+                        .takes_value(false)
+                        .help("Display balance in lamports instead of SOL"),
+                ),
+        )
+    }
+}
+
+pub fn parse_nonce_create_account(matches: &ArgMatches<'_>) -> Result<CliCommandInfo, CliError> {
+    let nonce_account = keypair_of(matches, "nonce_account_keypair").unwrap();
+    let lamports = amount_of(matches, "amount", "unit").expect("Invalid amount");
```
Comment (CriesofCarrots) [inline cli/src/nonce.rs]: I know we do this expect a few places in cli.rs and stake.rs, but can we `.ok_or()` to return an error here instead? 

Alternately, we could add a validator to the `lamports` Args to check they are parseable as ints, making this panic unreachable.

(I'll fixup the non-Nonce spots to follow your lead :) )

File: cli/src/nonce.rs [hunk 3]
```diff
@@ -0,0 +1,466 @@
...
+    if rpc_client.get_account(&nonce_account_pubkey).is_ok() {
+        return Err(CliError::BadParameter(format!(
+            "Unable to create nonce account. Nonce account already exists: {}",
+            nonce_account_pubkey
+        ))
+        .into());
+    }
+
+    let minimum_balance =
+        rpc_client.get_minimum_balance_for_rent_exemption(std::mem::size_of::<NonceState>())?;
```
Comment (t-nelson) [inline cli/src/nonce.rs]: Will `std::mem::size_of()` match `serialized_size()` here?  I had a conflict between the two at some point, hence `NonceState:size()`, which could be used here.

File: cli/src/nonce.rs [hunk 4]
```diff
@@ -0,0 +1,466 @@
...
+) -> ProcessResult {
+    let nonce_account = rpc_client.get_account(nonce_account_pubkey)?;
+    if nonce_account.owner != nonce_program::id() {
+        return Err(CliError::RpcRequestError(
+            format!("{:?} is not a nonce account", nonce_account_pubkey).to_string(),
+        )
+        .into());
+    }
+    match nonce_account.state() {
+        Ok(NonceState::Uninitialized) => Ok("Nonce account is uninitialized".to_string()),
```
Comment (t-nelson) [inline cli/src/nonce.rs]: An uninitialized account will still have a balance that the user may be interested in

File: cli/src/cli.rs [hunk 5]
```diff
@@ -1271,6 +1297,39 @@ pub fn process_command(config: &CliConfig) -> ProcessResult {
...
+        // Create nonce account
+        CliCommand::CreateNonceAccount {
+            nonce_account,
+            lamports,
+        } => process_create_nonce_account(&rpc_client, config, nonce_account, *lamports),
+        // Get a new nonce
+        CliCommand::NewNonce(nonce_account) => {
+            process_new_nonce(&rpc_client, config, nonce_account)
+        }
+        // Show the contents of a nonce account
```
Comment (t-nelson) [inline cli/src/cli.rs]: Maybe `// Retrieve nonce value stored in account` for this one?

File: cli/src/nonce.rs (hunk 2 above)
Comment (jackcmay) [reply cli/src/nonce.rs]: I added a validator :-)

**Comment by jackcmay** [comment]
@t-nelson Should also add support for `initialize`?

**Comment by t-nelson** [comment]
> @t-nelson What's the point of `initialize`? In what use cases do we expect to see uninitialized nonce accounts?

We will [soon](https://github.com/solana-labs/solana/issues/7305) need a mechanism for setting an "Authorized Noncer" (ala Authorized Staker) when an account is initialized.  Foisting this onto `Nonce` is a bit much, so `Initialize`.

We don't anticipate seeing an uninitialized nonce account as `create_nonce_account` will return both a `CreateAccount` and an `Initialize` instruction.  Since `initialize()` is now [private](https://github.com/solana-labs/solana/pull/7378/commits/8c75be1daa9325ec18ab450db8e59fe6ee6683d5), the user will have to go far out of their way for an account to exist in the `Uninitialized` state.  Though we should be prepared for this condition, IMO

**Pull request merged**